<a href="https://colab.research.google.com/github/W-B28/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/William_Bell_Assignment_LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

In [0]:
##### Begin Working Here #####
import pandas as pd


In [0]:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

df.head()

In [0]:
# display options

pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 200)

df.dtypes

In [0]:
# Look at the first ten rows of the `BedroomAbvGr` column.

df['BedroomAbvGr'].head()

In [0]:
df['LotFrontage'].head(10)

In [0]:
# counts of possible values in LotFrontage Column

df['LotFrontage'].value_counts(dropna=False)

In [0]:
import numpy as np

type(np.NaN)

### Making New Features

In [0]:
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']]

small_df

In [0]:
# Create a new column

small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF'] 

small_df

## Load new dataset

In [0]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

LoanStats_2018Q4.cs     [                 <=>]  21.72M   814KB/s    in 28s     

2020-01-15 16:14:53 (796 KB/s) - ‘LoanStats_2018Q4.csv.zip.1’ saved [22776791]



In [0]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
replace LoanStats_2018Q4.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


In [0]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [0]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

In [0]:
df = pd.read_csv('LoanStats_2018Q4.csv')
print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128400, 1)


Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id  member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title           emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url desc purpose            title                   zip_code addr_state dti   delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt     total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
NaN NaN       6000      6000        6000             36 months  14.47%  206.44      C     C2        Manager             10+ years  RENT           55000      Verified            Dec-2018 Charged Off n          NaN NaN  credit_card        Credit card refinancing 104xx    NY         10.65 0           Jun-2006         0              NaN                    NaN                    6        0       4196      91.2%      8         w                   0.00      0.00          1106.52         1106.52         546.14          272.38        0.0                288.0      51.84                   Apr-2019     206.44          NaN          Nov-2019           0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            12514       0           2           1           1           9                  8318         74      1           1           2807       79       4600             1      0           2            2                    2086        344            92      0                        0           33                 150                  9                     9              0        9                    NaN                      9                     NaN                            0                     3              4               3           3         3         4             5             4                   6        0                0            0                  2                  100            100        

^^^ How did we know those NaN were coming from the bottom two rows?

In [0]:
# We can fix the header problem by using the 'skiprows' or 'header' parameters

df = pd.read_csv('LoanStats_2018Q4.csv', header=1)
print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128399, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65          0.0         Jun-2006             0.0   
1         IL  19.14          0.0         Oct-1996             3.0   
2         NE  22.23          0.0         May-2007             1.0   
3         MI   9.44          0.0         Mar-2009             2.0   
4         VA  18.47          0.0         Dec-2004             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       6.0      0.0   
1                    39.0                     NaN      14.0      0.0   
2                     NaN                     NaN       8.0      0.0   
3                    73.0                     NaN       7.0      0.0   
4                    75.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0     4196.0      91.2%        8.0                   w       0.00   
1     8547.0      27.7%       28.0                   w   16807.30   
2    25754.0        48%       23.0                   w       0.00   
3    11434.0      33.3%       20.0                   w    1707.34   
4    13555.0      38.2%       22.0                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymn

^^^ How did the shape change from (128400, 1) to (128399, 144) by changing the header

---



In [0]:
df = pd.read_csv('LoanStats_2018Q4.csv', header=1 )

print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128399, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65          0.0         Jun-2006             0.0   
1         IL  19.14          0.0         Oct-1996             3.0   
2         NE  22.23          0.0         May-2007             1.0   
3         MI   9.44          0.0         Mar-2009             2.0   
4         VA  18.47          0.0         Dec-2004             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       6.0      0.0   
1                    39.0                     NaN      14.0      0.0   
2                     NaN                     NaN       8.0      0.0   
3                    73.0                     NaN       7.0      0.0   
4                    75.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0     4196.0      91.2%        8.0                   w       0.00   
1     8547.0      27.7%       28.0                   w   16807.30   
2    25754.0        48%       23.0                   w       0.00   
3    11434.0      33.3%       20.0                   w    1707.34   
4    13555.0      38.2%       22.0                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymn

In [0]:

df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine='python')
print(df.shape)
df.head()

(128397, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       6000         6000           6000.0   36 months   
1 NaN        NaN      24000        24000          24000.0   36 months   
2 NaN        NaN      30000        30000          30000.0   36 months   
3 NaN        NaN       8000         8000           8000.0   36 months   
4 NaN        NaN      22400        22400          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65            0         Jun-2006               0   
1         IL  19.14            0         Oct-1996               3   
2         NE  22.23            0         May-2007               1   
3         MI   9.44            0         Mar-2009               2   
4         VA  18.47            0         Dec-2004               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN         6        0   
1                    39.0                     NaN        14        0   
2                     NaN                     NaN         8        0   
3                    73.0                     NaN         7        0   
4                    75.0                     NaN        15        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       4196      91.2%          8                   w       0.00   
1       8547      27.7%         28                   w   16807.30   
2      25754        48%         23                   w       0.00   
3      11434      33.3%         20                   w    1707.34   
4      13555      38.2%         22                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymnt_d  

In [0]:
df.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
128392 NaN        NaN      23000        23000          23000.0   36 months   
128393 NaN        NaN      10000        10000          10000.0   36 months   
128394 NaN        NaN       5000         5000           5000.0   36 months   
128395 NaN        NaN      10000        10000           9750.0   36 months   
128396 NaN        NaN      10000        10000          10000.0   36 months   

       int_rate  installment grade sub_grade             emp_title emp_length  \
128392   15.02%       797.53     C        C3        Tax Consultant  10+ years   
128393   15.02%       346.76     C        C3        security guard    5 years   
128394   13.56%       169.83     C        C1          Payoff Clerk  10+ years   
128395   11.06%       327.68     B        B3                   NaN        NaN   
128396   16.91%       356.08     C        C5  Key Accounts Manager    2 years   

       home_ownership  annual_inc verification_status   issue_d  loan_status  \
128392       MORTGAGE     75000.0     Source Verified  Oct-2018  Charged Off   
128393       MORTGAGE     38000.0        Not Verified  Oct-2018      Current   
128394       MORTGAGE     35360.0        Not Verified  Oct-2018      Current   
128395           RENT     44400.0     Source Verified  Oct-2018      Current   
128396           RENT     80000.0        Not Verified  Oct-2018      Current   

       pymnt_plan  url  desc             purpose                    title  \
128392          n  NaN   NaN  debt_consolidation       Debt consolidation   
128393          n  NaN   NaN  debt_consolidation       Debt consolidation   
128394          n  NaN   NaN  debt_consolidation       Debt consolidation   
128395          n  NaN   NaN         credit_card  Credit card refinancing   
128396          n  NaN   NaN               other                    Other   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128392    352xx         AL  20.95            1         Aug-1985   
128393    443xx         OH  13.16            3         Jul-1982   
128394    381xx         TN  11.30            1         Jun-2006   
128395    980xx         WA  11.78            0         Oct-2008   
128396    021xx         MA  17.72            1         Sep-2006   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128392               2                    22.0                     NaN   
128393               0                     6.0                     NaN   
128394               0                    21.0                     NaN   
128395               2                    40.0                     NaN   
128396               0                    14.0                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128392        12        0      22465      43.6%         28   
128393        11        0       5634      37.1%         16   
128394         9        0       2597      27.3%         15   
128395        15        0       6269      13.1%         25   
128396        17        0       1942      30.8%         31   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128392                   w       0.00           0.00      1547.08   
128393                   w    6632.21        6632.21      4846.30   
128394                   f    3291.41        3291.41      2373.85   
128395                   f    6497.81        6335.37      4572.16   
128396                   w    6695.63        6695.63      4975.73   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
128392          1547.08          1025.67         521.41                 0.0   
128393          4846.30          3367.79        1478.51                 0.0   
128394          2373.85          1708.59         665.26                 0.0   
128395          4457.86          3502.19        1069.97                 0.0   
128396          4975.73          3304.37        1671.36                 0.0  

In [0]:
# Sum null values by column and sort from least to greatest
pd.set_option('display.max_rows', 200)

df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
hardship_flag                                      0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

In [0]:
# dropping the empty columns 

df.shape

# df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

# df.head()


(128397, 144)

In [0]:
### Look at the first 10 values of the int_rate column
df['int_rate'][:10]

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

In [0]:
### Look at a specific value from the int_rate column

int_rate = df['int_rate'][2]

int_rate


' 15.02%'

In [0]:
"""
Problems that we need to address with this column:

String column that should be numeric
Percent Sign % included with the number
Leading space at the beginning of the string

"""


print(int_rate.strip().strip('%'))

type(int_rate.strip().strip('%'))

15.02


str

In [0]:
### "Cast" the string value to a float

float(int_rate.strip().strip('%'))

15.02

In [0]:
type(float(int_rate.strip().strip('%')))

float

In [0]:
### Write a function that can do what we have written above to any 
### string that is passsed to it.
term_rate = df['term'][2]


def str_to_int(cell_contents):
  return int(cell_contents.strip().strip(' months'))

str_to_int(term_rate)

36

In [0]:
### is the data type correct?

type(str_to_int(term_rate))

int

In [84]:
# .apply()to every cell in term column from str to int

df['term_integer'] = df['term_rate'].apply(str_to_int)

df.head()

KeyError: ignored

In [91]:
# Make a column named 'loan_status_is_great'. It should contain the integer 1 
# if 'loan_status' is "Current" or "Fully Paid." Else it should contain the integer 0.

df['loan_status_is_great'] = df[df['loan_status'] == "Current" or df['loan_status'] == "Fully Paid"]


def str_to_int(loan_status):
  return 1 if loan_status is True else 0

df['loan_status_is_great'] = df['loan_status_is_great'].apply(str_to_int)

ValueError: ignored

In [95]:
# Make last_pymnt_d_month and last_pymnt_d_year columns.

import pandas as pd

[col for col in df if col.endswith('_d')]

df['last_pymnt_d'] = pd.to_datetime(df['last_pymnt_d'], infer_datetime_format=True)

df['last_pymnt_d'].head().values


array(['2019-04-01T00:00:00.000000000', '2019-12-01T00:00:00.000000000',
       '2019-02-01T00:00:00.000000000', '2019-12-01T00:00:00.000000000',
       '2019-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [96]:
df['last_pymnt_d_month'] = df['last_pymnt_d'].dt.month

df['last_pymnt_d_month']

0          4.0
1         12.0
2          2.0
3         12.0
4         12.0
          ... 
128392    12.0
128393    12.0
128394    12.0
128395    12.0
128396    12.0
Name: last_pymnt_d_month, Length: 128397, dtype: float64

In [97]:
df['last_pymnt_d_year'] = df['last_pymnt_d'].dt.year

df['last_pymnt_d_year']

0         2019.0
1         2019.0
2         2019.0
3         2019.0
4         2019.0
           ...  
128392    2018.0
128393    2019.0
128394    2019.0
128395    2019.0
128396    2019.0
Name: last_pymnt_d_year, Length: 128397, dtype: float64

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

You can uncomment and run the cells below to re-download and extract the Instacart data

In [0]:
# !wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# !tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# %cd instacart_2017_05_01